File: VB/GridControlTypeConverter/CustomData/TypeConverterHelper.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Linq
Imports System.Text
Imports System.Windows.Forms
Imports DevExpress.XtraEditors.Repository
Imports DevExpress.XtraGrid.Views.Base
Imports DevExpress.XtraGrid.Views.Grid
Imports DevExpress.XtraGrid.Columns
Imports System.Collections

Namespace GridControlTypeConverter
	Public Class TypeConverterHelper
		Private colView As GridView
		Private unboundColumnFieldName As String = "UnboundColumn", convertedProperty As String = String.Empty
		Public Sub New(ByVal unboundColumnFieldName As String, ByVal convertedProperty As String, ByVal view As GridView)

			Me.unboundColumnFieldName = unboundColumnFieldName
			Me.convertedProperty = convertedProperty
			colView = view
			CreateUnboundColumn(view, unboundColumnFieldName)
		End Sub
		Private Sub SubscribeToEvents()
			AddHandler colView.CustomUnboundColumnData, AddressOf OnCustomUnboundColumnData
		End Sub
		Private Sub CreateUnboundColumn(ByVal view As GridView, ByVal unboundColumnFieldName As String)

			view.PopulateColumns()
			Dim originColumn As GridColumn = view.Columns(convertedProperty)
			If originColumn IsNot Nothing Then
				Dim column As GridColumn = view.Columns.AddVisible(unboundColumnFieldName)
				column.UnboundType = DevExpress.Data.UnboundColumnType.Object
				SubscribeToEvents()
			End If
		End Sub
		Private Sub OnCustomUnboundColumnData(ByVal sender As Object, ByVal e As DevExpress.XtraGrid.Views.Base.CustomColumnDataEventArgs)
			If e.Column.FieldName = unboundColumnFieldName Then
				If e.IsGetData Then
					Dim dataSource As IList = TryCast(colView.DataSource, IList)
					Dim obj As Object = dataSource(e.ListSourceRowIndex)
					If obj Is Nothing Then
						Return
					End If
					Dim descriptor As PropertyDescriptor = TypeDescriptor.GetProperties(obj.GetType())(convertedProperty)
					If descriptor Is Nothing Then
						Return
					End If
					Dim value As Object = descriptor.GetValue(obj)
					Dim converter As TypeConverter = descriptor.Converter
					If converter IsNot Nothing AndAlso converter.CanConvertFrom(value.GetType()) AndAlso TypeOf value Is System.Enum Then
						e.Value = converter.ConvertFrom(value.ToString())
					ElseIf converter IsNot Nothing AndAlso converter.CanConvertFrom(value.GetType()) Then
						e.Value = converter.ConvertFrom(value)
					End If
				ElseIf e.IsSetData Then
					Dim dataSource As IList = TryCast(colView.DataSource, IList)
					Dim obj As Object = dataSource(e.ListSourceRowIndex)
					If obj Is Nothing Then
						Return
					End If
					Dim descriptor As PropertyDescriptor = TypeDescriptor.GetProperties(obj.GetType())(convertedProperty)
					If descriptor Is Nothing Then
						Return
					End If
					Dim converter As TypeConverter = descriptor.Converter
					Dim value As Object = descriptor.GetValue(obj)
					If converter IsNot Nothing AndAlso converter.CanConvertTo(value.GetType()) Then
						descriptor.SetValue(obj, converter.ConvertTo(e.Value, value.GetType()))
					End If
				End If
			End If
		End Sub
		Private Sub UnsubcribeFromEvents()
			RemoveHandler colView.CustomUnboundColumnData, AddressOf OnCustomUnboundColumnData
		End Sub

		Public Sub RemoveUnboundColumn()
			Dim column As GridColumn = colView.Columns(unboundColumnFieldName)
			Dim originColumn As GridColumn = colView.Columns(convertedProperty)
			If column IsNot Nothing AndAlso originColumn IsNot Nothing Then
				colView.Columns.Remove(column)
				originColumn.Visible = True
				UnsubcribeFromEvents()
			End If
		End Sub
	End Class
End Namespace

File: CS/GridControlTypeConverter/CustomData/TypeConverterHelper.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using DevExpress.XtraEditors.Repository;
using DevExpress.XtraGrid.Views.Base;
using DevExpress.XtraGrid.Views.Grid;
using DevExpress.XtraGrid.Columns;
using System.Collections;

namespace GridControlTypeConverter
{
    public class TypeConverterHelper
    {       
        GridView colView;
        string unboundColumnFieldName = "UnboundColumn",
            convertedProperty = string.Empty;     
        public TypeConverterHelper(string unboundColumnFieldName,
            string convertedProperty, GridView view)
        {
            
            this.unboundColumnFieldName = unboundColumnFieldName;
            this.convertedProperty = convertedProperty;
            colView = view;
            CreateUnboundColumn(view, unboundColumnFieldName);          
        }
        void SubscribeToEvents()
        {
            colView.CustomUnboundColumnData += OnCustomUnboundColumnData;
        }
        private void CreateUnboundColumn(GridView view, string unboundColumnFieldName)
        {

            view.PopulateColumns();
            GridColumn originColumn = view.Columns[convertedProperty];
            if (originColumn != null)
            {             
                GridColumn column = view.Columns.AddVisible(unboundColumnFieldName);
                column.UnboundType = DevExpress.Data.UnboundColumnType.Object;         
                SubscribeToEvents();
            }
        }
        void OnCustomUnboundColumnData(object sender, DevExpress.XtraGrid.Views.Base.CustomColumnDataEventArgs e)
        {
            if (e.Column.FieldName == unboundColumnFieldName)
            {
                if (e.IsGetData)
                {
                    IList dataSource = colView.DataSource as IList;
                    object obj = dataSource[e.ListSourceRowIndex];
                    if (obj == null) return;
                    PropertyDescriptor descriptor = TypeDescriptor.GetProperties(obj.GetType())[convertedProperty];
                    if (descriptor == null) return;
                    object value = descriptor.GetValue(obj);
                    TypeConverter converter = descriptor.Converter;
                    if (converter != null && converter.CanConvertFrom(value.GetType()) && value is Enum)
                        e.Value = converter.ConvertFrom(value.ToString());
                    else if (converter != null && converter.CanConvertFrom(value.GetType()))
                    {
                        e.Value = converter.ConvertFrom(value);                      
                    }
                }
                else if (e.IsSetData)
                {
                    IList dataSource = colView.DataSource as IList;
                    object obj = dataSource[e.ListSourceRowIndex];
                    if (obj == null) return;
                    PropertyDescriptor descriptor = TypeDescriptor.GetProperties(obj.GetType())[convertedProperty];
                    if (descriptor == null) return;                        
                    TypeConverter converter = descriptor.Converter;
                    object value = descriptor.GetValue(obj);
                    if (converter != null && converter.CanConvertTo(value.GetType()))
                        descriptor.SetValue(obj, converter.ConvertTo(e.Value, value.GetType()));
                }
            }
        }
        void UnsubcribeFromEvents()
        {
            colView.CustomUnboundColumnData -= OnCustomUnboundColumnData;
        }

        public void RemoveUnboundColumn()
        {
            GridColumn column = colView.Columns[unboundColumnFieldName];
            GridColumn originColumn = colView.Columns[convertedProperty];
            if (column != null && originColumn != null)
            {
                colView.Columns.Remove(column);
                originColumn.Visible = true;
                UnsubcribeFromEvents();
            }
        }
    }
}
